### ISSUE TITLE
[Discussion][Feature] Clean up workersToDelete?

### ISSUE THREAD

**Issue by DmitriGekhtman** [ray-project/kuberay#733]
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/kuberay/issues) and found no similar feature requirement.

### Description

We might want the operator to clean up workersToDelete after it is done deleting the requested workers.
Currently, if a user sets workersToDelete, the autoscaler will keep trying to delete the workers, even after they are gone. This leads to some confusing log messages.

## Pros

- Avoid confusing logs. (This can be achieved in simpler ways, of course.)
- Automatically clean up irrelevant configuration.
- Potentially, act as an ack to user/autoscaler that the the operator has done its job in deleting pods.
- Similar controllers do this: https://github.com/openkruise/kruise/blob/a639a6f55721bf2fc5254b0d4499ae94ef9a1851/pkg/controller/cloneset/cloneset_controller.go#L328

## Cons

- Increases complexity of the code.
- The user can clean up the workersToDelete field themselves.
- It is not idiomatic for an operator to write to the spec of the resource it manages -- in fact, that can lead to annoying 409 conflicts due to optimistic concurrency control.

### Use case

Achieve the pros above.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by DmitriGekhtman** [comment]
cc @kevin85421

**Comment by lenliu3** [comment]
Seeing a problem related to this with a reproducible pattern. Adding details in case it helps prioritize a fix.

## Environment

- KubeRay operator: helm chart `ray-cluster-1.5.1`
- Ray: `2.53.0`
- Autoscaler: v2 (in-tree)
- Kubernetes: EKS v1.31.10

## Configuration

```yaml
enableInTreeAutoscaling: true
autoscalerOptions:
  version: v2
  idleTimeoutSeconds: 0
workerGroupSpecs:
  - groupName: pool
    replicas: 0
    minReplicas: 0
    maxReplicas: 100
    resources:
      requests: { cpu: "4", memory: "8Gi" }
      limits: { cpu: "4", memory: "8Gi" }
```

## Workload

We submit batches of ~25 short-lived RayJobs (1-2 min each) via `K8sJobMode` against a persistent shared RayCluster. Batches arrive periodically throughout the day.

## What happens

After a batch completes and the cluster scales to zero, subsequent batches deadlock — no workers are ever created. All new jobs sit in `PENDING` for 900s then fail:

```
Job supervisor actor failed to start within 900.0 seconds.
```

`ray status` shows the autoscaler *thinks* it's launching workers, but the operator never creates them:

```
Pending:
 pool, 1 launching
Recent failures:
 pool: LaunchFailed (latest_attempt: 11:17:00)
 ...
Pending Demands:
 {'CPU': 0.1}: 12+ pending tasks/actors
```

## Root cause

The live RayCluster spec shows a stale `workersToDelete` referencing a pod that no longer exists:

```json
{
  "replicas": 0,
  "scaleStrategy": {
    "workersToDelete": ["pidp-ray-cluster-pool-worker-ddd67"]
  }
}
```

That pod was already deleted during a previous scale-down. The combination of `replicas: 0` + orphaned `workersToDelete` prevents the operator from creating new workers, even though the autoscaler is requesting them through GCS.

## Workaround

```bash
# Clear stale entries
kubectl patch raycluster <name> -n <ns> --type=json \
  -p '[{"op":"replace","path":"/spec/workerGroupSpecs/0/scaleStrategy/workersToDelete","value":[]}]'

# Kick replicas to restart autoscaler loop
kubectl patch raycluster <name> -n <ns> --type=json \
  -p '[{"op":"replace","path":"/spec/workerGroupSpecs/0/replicas","value":1}]'
```

After this, workers spin up immediately and jobs proceed normally.

**Comment by fangyinc** [comment]
> Seeing a problem related to this with a reproducible pattern. Adding details in case it helps prioritize a fix.
> 
> ## Environment
> * KubeRay operator: helm chart `ray-cluster-1.5.1`
> * Ray: `2.53.0`
> * Autoscaler: v2 (in-tree)
> * Kubernetes: EKS v1.31.10
> 
> ## Configuration
> enableInTreeAutoscaling: true
> autoscalerOptions:
>   version: v2
>   idleTimeoutSeconds: 0
> workerGroupSpecs:
>   - groupName: pool
>     replicas: 0
>     minReplicas: 0
>     maxReplicas: 100
>     resources:
>       requests: { cpu: "4", memory: "8Gi" }
>       limits: { cpu: "4", memory: "8Gi" }
> ## Workload
> We submit batches of ~25 short-lived RayJobs (1-2 min each) via `K8sJobMode` against a persistent shared RayCluster. Batches arrive periodically throughout the day.
> 
> ## What happens
> After a batch completes and the cluster scales to zero, subsequent batches deadlock — no workers are ever created. All new jobs sit in `PENDING` for 900s then fail:
> 
> ```
> Job supervisor actor failed to start within 900.0 seconds.
> ```
> 
> `ray status` shows the autoscaler _thinks_ it's launching workers, but the operator never creates them:
> 
> ```
> Pending:
>  pool, 1 launching
> Recent failures:
>  pool: LaunchFailed (latest_attempt: 11:17:00)
>  ...
> Pending Demands:
>  {'CPU': 0.1}: 12+ pending tasks/actors
> ```
> 
> ## Root cause
> The live RayCluster spec shows a stale `workersToDelete` referencing a pod that no longer exists:
> 
> {
>   "replicas": 0,
>   "scaleStrategy": {
>     "workersToDelete": ["pidp-ray-cluster-pool-worker-ddd67"]
>   }
> }
> That pod was already deleted during a previous scale-down. The combination of `replicas: 0` + orphaned `workersToDelete` prevents the operator from creating new workers, even though the autoscaler is requesting them through GCS.
> 
> ## Workaround
> # Clear stale entries
> kubectl patch raycluster <name> -n <ns> --type=json \
>   -p '[{"op":"replace","path":"/spec/workerGroupSpecs/0/scaleStrategy/workersToDelete","value":[]}]'
> 
> # Kick replicas to restart autoscaler loop
> kubectl patch raycluster <name> -n <ns> --type=json \
>   -p '[{"op":"replace","path":"/spec/workerGroupSpecs/0/replicas","value":1}]'
> After this, workers spin up immediately and jobs proceed normally.

Hi, @lenliu3 , has your issue been fixed? I seem to be encountering a similar problem.

**Comment by lenliu3** [comment]
@fangyinc I dont believe this is fixed. Still seeing a periodic increase of stale workersToDelete build up over time on our side which requires manual patching to clear the list.
